Document content of the standard library (#517)
openqasm/openqasm: source/language/standard_library.rst

Standard library
================

The standard library for OpenQASM 3 consists of a single include file, called ``stdgates.inc``.
This file can be included in an OpenQASM 3 program with the statement

.. code-block::

   include "stdgates.inc";

Versioning
----------

The standard library is versioned directly with the language itself.  This documentation will note
the version that any given gate in the standard library became available, and, if applicable, the
version it was removed in.

Notes for implementations
-------------------------

An implementation is not required to make the standard library available in situations where the
OpenQASM 3 program is being interpreted in terms of the instruction set architecture of a particular
QPU.  The rest of this section concerns only situations where the library is considered available.

Implementations should consider the ``stdgates.inc`` file to be locatable, regardless of any other
settings that they might apply around include search paths.

Upon interpreting an ``include "stdgates.inc";`` statement, implementations should make available

> exactly the set of gates defined in the standard library for the language version that matches the

Author: There'd been some worry from people that if the file had versions, it would be a bit of a nuance to manage if all implementations actually had to parse a real file backing it, since the compatibility version of the parse might only be determined dynamically when the `OPENQASM 3.x` statement is encountered..  The point of the provision was to make it clear that as long as an implementation behaves _as if_ there was a real file backing it, that's fine.

:ref:`version-string statement <version-string>`, if present.  If the version statement is not
present, the implementation should make available the gates from the language version it is parsing
the rest of the program under.

> Implementations must not define gates that are not present in chosen OpenQASM 3 version, where

Reviewer: ```suggestion
Implementations must not define gates that are not present in the chosen OpenQASM 3 version, where
```

> doing so could cause a user program to be ill formed (for example due to a user attempting to define

Reviewer: Wouldn't *any* other gate definition potentially (silently) conflict with a user's gate definition?

their own gate with a clashing name).

> An implementation is not required to supply a literal ``stdgates.inc`` file to the user.

Reviewer 2: Should we update https://openqasm.com/language/comments.html#included-files in any way?

Author: What kind of thing were you thinking?  We should probably put in a link to this page as a cross-reference to the `stdlib.inc` file for certain.

Author: I've included some cross-referencing to other places in 885ab15 - happy to add more.

Reviewer 2: Looks good, thank you.

   Equivalent to ``ctrl @ gphase(λ) a``.

   .. seealso::

      :gate:`rz`
         The same operation up to a global phase.

      :gate:`phase`
         An alternative name for backwards compatibility with OpenQASM 2.

   .. versionadded:: 3.0

.. gate:: x a

   The Pauli X gate.  Defined by the mapping :math:`\lvert0\rangle \leftrightarrow \lvert1\rangle`.

   .. versionadded:: 3.0

.. gate:: y a

   The Pauli Y gate.  Defined by the mapping :math:`\lvert0\rangle \to i\lvert1\rangle` and
   :math:`\lvert1\rangle \to -i\lvert0\rangle`.

   .. versionadded:: 3.0

.. gate:: z a

   The Pauli Z gate.  Defined by the mapping :math:`\lvert0\rangle \to \lvert0\rangle` and
   :math:`\lvert1\rangle \to \lvert1\rangle`.

   Equivalent to ``p(pi) a``.

   .. versionadded:: 3.0

.. gate:: h a

   The Hadamard gate.  Defined by the mapping
   :math:`\lvert0\rangle \to \bigl(\lvert0\rangle + \lvert1\rangle\bigr)/\sqrt2` and
   :math:`\lvert1\rangle \to \bigl(\lvert0\rangle - \lvert1\rangle\bigr)/\sqrt2`.

   .. versionadded:: 3.0

.. gate:: s a

   The :math:`\sqrt Z` gate (see :gate:`z`).  The square root is chosen conventionally, that is

>    the gate is equivalent to :math:`P(\pi/2)`, in terms of :gate:`p`.

Author: I had deliberately used the mathematical gate symbols, then linked them to the associated OpenQASM gate, to try and make them understandable at a glance to people familiar with the maths, but I've no problem accepting changes on it if people don't think it's clear.

>    the gate is equivalent to :math:`P(\pi/2)`, in terms of :gate:`p`.

Reviewer 3: I don't mind too much, but I did find myself wondering what `P` was and whether it was supposed to be the gate `p`.

>    the gate is equivalent to :math:`P(\pi/2)`, in terms of :gate:`p`.

Reviewer 3: Maybe my real question is where do I find the mathematical definition of `P`? Is there a standard definition?

>    the gate is equivalent to :math:`P(\pi/2)`, in terms of :gate:`p`.

Author: To avoid all doubt, I've pushed 42fd895 which includes formal mathematics and symbols for each gate.  I've deliberately avoided using matrix representations, and I was very explicit about the tensor indices of each qubit in each example.

>    the gate is equivalent to :math:`P(\pi/2)`, in terms of :gate:`p`.

Reviewer 3: Thanks @jakelishman. The formal basis transformation definitions are perhaps overkill but definitely cool and resolves the uncertainty I was worried about.

   the gate is equivalent to :math:`P(\pi/4)`, in terms of :gate:`p`.

   .. versionadded:: 3.0

.. gate:: tdg a

   Adjoint of :gate:`t`.  Equivalent to :math:`P(-\pi/4)`, in terms of :gate:`p`.

   .. versionadded:: 3.0

.. gate:: sx a

   The :math:`\sqrt X` gate (see :gate:`x`).

   Explicitly, this has the action
   :math:`\lvert0\rangle \to \Bigl((1 + i)\lvert0\rangle + (1 - i)\lvert1\rangle\Bigr)/2` and
   :math:`\lvert1\rangle \to \Bigl((1 - i)\lvert0\rangle + (1 + i)\lvert1\rangle\Bigr)/2`.

   .. versionadded:: 3.0

.. gate:: rx(θ) a

   Rotation about the :math:`X` axis: :math:`\exp(-i\theta X)`.

   .. versionadded:: 3.0

.. gate:: ry(θ) a

   Rotation about the :math:`Y` axis: :math:`\exp(-i\theta Y)`.

   .. versionadded:: 3.0

.. gate:: rz(θ) a

   Rotation about the :math:`Z` axis: :math:`\exp(-i\theta Z)`.  Note that this differs from
   :gate:`p` by a global phase of half the rotation angle.

   .. seealso::
      :gate:`p`
         The same gate but with a different global-phase convention.

   .. versionadded:: 3.0

Two-qubit gates
...............

All of the controlled gates defined in the standard library follow the same conventions of the
``ctrl`` modifier.

> Explicitly, the first qubit is the control and the second the target.

Reviewer: I was taught, and still follow, using "in particular" rather than "explicitly" here. There are an unspecified number of "same conventions". Using "in particular" (especially if it is understood as a math/science idiom) says that we are drawing attention to one particular item as especially important in the present context.

I might be accused of proscriptivism, but I think it is sometimes warranted.

Author: The controlled gates have all the same conventions as the controlled modifier (it just happens that there's only really one), so I think "in particular" would be misleading - we're not selecting one of many, we're linking together and then being explicit about what those linked conventions are.

Reviewer: Ok. This is clear.

OpenQASM 3 program is being interpreted in terms of the instruction set architecture of a particular
QPU.  The rest of this section concerns only situations where the library is considered available.

Implementations should consider the ``stdgates.inc`` file to be locatable, regardless of any other
settings that they might apply around include search paths.

Upon interpreting an ``include "stdgates.inc";`` statement, implementations should make available
exactly the set of gates defined in the standard library for the language version that matches the
:ref:`version-string statement <version-string>`, if present.  If the version statement is not
present, the implementation should make available the gates from the language version it is parsing
the rest of the program under.

Implementations must not define gates that are not present in the chosen OpenQASM 3 version, where
doing so could cause a user program to be ill formed (for example due to a user attempting to define
their own gate with a clashing name).

An implementation is not required to supply a literal ``stdgates.inc`` file to the user.

> An implementation must make all the gates available as if they were defined with ``gate``
> statements with contents that match the mathematical descriptions given here, but it is explicitly
> permissible for implementations to internally treat the ``stdgates.inc`` include as a series of
> ``defcal`` statements for all qubits used in the program.

Reviewer 3: ```suggestion
An implementation must make all the gates available as if they were defined with ``gate``
statements with contents that match the mathematical descriptions given here.
In addition, it is permissible for implementations to also provide implementations of the gates
equivalent to a series of ``defcal`` statements for all qubits used in the program.
```
I wanted to clarify that the implementations are in addition to the gate definitions.

   or explicitly:

   .. math::

      \texttt{cry(θ) a, b;} \mathit{CRY}_{ba}(\theta)\colon\ \left\{\begin{aligned}[c]
         {\lvert00\rangle}_{ba} &\to {\lvert00\rangle}_{ba}\\
         {\lvert01\rangle}_{ba} &\to {\bigl(\cos\theta\lvert0\rangle + \sin\theta\lvert1\rangle\bigr)}_b{\lvert1\rangle}_a\\
         {\lvert10\rangle}_{ba} &\to {\lvert10\rangle}_{ba}&\\
         {\lvert11\rangle}_{ba} &\to {\bigl(\cos\theta\lvert1\rangle - \sin\theta\lvert0\rangle\bigr)}_b{\lvert1\rangle}_a\\
      \end{aligned}\right.

   .. versionadded:: 3.0

.. gate:: crz(θ) a, b

   Controlled :math:`Z` rotation with an angle :math:`\theta` (see :gate:`rz`). Its mapping is defined by
   :math:`\mathit{CRZ}_{ba}(\theta) = I_b{\lvert0\rangle\!\langle0\rvert}_a + \mathit{CRZ}_b(\theta){\lvert1\rangle\!\langle1\rvert}_a`,
   or explicitly:

   .. math::

      \texttt{crz(θ) a, b;} \mathit{CRZ}_{ba}(\theta)\colon\ \left\{\begin{alignedat}[c]2
         {\lvert00\rangle}_{ba} &{}\to{}& &{\lvert00\rangle}_{ba}\\
         {\lvert01\rangle}_{ba} &{}\to{}& e^{-i\theta/2} &{\lvert01\rangle}_{ba}\\

>          {\lvert10\rangle}_{ba} &{}\to{}& &{\lvert10\rangle}_{ba} \vphantom{e^{i\theta/2}}\\

Author: MathJax handles it find.  I was just being silly about the line spacings all coming out looking the same...

>          {\lvert10\rangle}_{ba} &{}\to{}& &{\lvert10\rangle}_{ba} \vphantom{e^{i\theta/2}}\\

Author: <img width="699" alt="Screenshot 2024-04-17 at 16 25 40" src="https://github.com/openqasm/openqasm/assets/5968590/edd2dac0-c6fc-432d-99ee-3dd4dc47afee">

Sample rendering

>          {\lvert10\rangle}_{ba} &{}\to{}& &{\lvert10\rangle}_{ba} \vphantom{e^{i\theta/2}}\\

Author: ... and that picture just made me realise there's a typo in the inlined version of the CRX action.
